aspects/aspects.go

```diff
@@ -221,13 +230,82 @@ func (a *Aspect) Get(name string, value interface{}) error {
 	return &NotFoundError{fmt.Sprintf("cannot get %q: name not found", name)}
 }
 
-// accessPattern holds information on how to access an aspect.
+func newAccessPattern(name, path, accesstype string) (*accessPattern, error) {
+	accType, err := newAccessType(accesstype)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create aspect pattern: %w", err)
+	}
+
+	split := func(str string) []pattern {
+		parts := strings.Split(str, ".")
+		patterns := make([]pattern, 0, len(parts))
+		for _, p := range parts {
+			patterns = append(patterns, pattern(p))
+		}
+		return patterns
+	}
+
+	return &accessPattern{
+		name:   split(name),
+		path:   split(path),
+		access: accType,
+	}, nil
+}
+
+// accessPattern represents an individual aspect access pattern. It can be used
+// to match an input name and map it into a corresponding path, potentially with
+// placeholders filled in.
 type accessPattern struct {
-	name   string
-	path   string
+	name   []pattern
+	path   []pattern
 	access accessType
 }
 
+// match takes a list of parts and returns true if those parts match the pattern's
+// name. If the name contains placeholders, those will be mapped to their values in
+// the supplied parts and set in the map. Example: if pattern.name=["{foo}", "b", "{bar}"],
+// and nameParts=["a", "b", "c"], then it returns true and the map will contain
+// {"foo": "a", "bar": "c"}.
+func (p *accessPattern) match(nameParts []string, placeholders map[string]string) bool {
+	if len(p.name) != len(nameParts) {
+		return false
+	}
+
+	for i, part := range nameParts {
+		if !p.name[i].match(part, placeholders) {
+			// clearing the map isn't strictly necessary because a path's placeholders
+			// must be in the name so they would be overwritten but let's be robust
+			for k := range placeholders {
+				delete(placeholders, k)
+			}
```

Comment:
this is strange, it seems like it make morse for match at the accessPattern level to return the placeholders instead of taking the map, while taking the make sense for the pattern structs

Reply:
It was a premature optimisation I guess, so we wouldn't reallocate it each time. I'll remove it